core/libp2p/connection/yamux/yamuxed_connection.cpp

```diff
@@ -0,0 +1,494 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/connection/yamux/yamuxed_connection.hpp"
+
+#include "libp2p/connection/yamux/yamux_frame.hpp"
+#include "libp2p/connection/yamux/yamux_stream.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(libp2p::connection, YamuxedConnection::Error, e) {
+  using ErrorType = libp2p::connection::YamuxedConnection::Error;
+  switch (e) {
+    case ErrorType::NO_SUCH_STREAM:
+      return "no such stream was found; maybe, it is closed";
+    case ErrorType::YAMUX_IS_CLOSED:
+      return "this Yamux instance is closed";
+    case ErrorType::FORBIDDEN_CALL:
+      return "forbidden method was invoked";
+    case ErrorType::OTHER_SIDE_ERROR:
+      return "error happened on other side's behalf";
+    case ErrorType::INTERNAL_ERROR:
+      return "internal error happened";
+  }
+  return "unknown";
+}
+
+#define HANDLE_WRITE_ERROR                               \
```

Comment:
Why not an inline function? It's just that macros are generally less reliable so without an imminent need I would prefer not to have them in code.